### ISSUE TITLE
Error: Payload Limit Exceeded

### ISSUE THREAD

**Issue by arnavroh45** [qdrant/qdrant-client#463]
I am trying to store LlamaIndex Documents in the qdrant database(docker). When I try storing them in the db. I am getting this error. Please help me solve this.

UnexpectedResponse: Unexpected Response: 400 (Bad Request)
Raw response content:
b'{"status":{"error":"Payload error: JSON payload (46866880 bytes) is larger than allowed (limit: 33554432 bytes)."},"time":0.0}'

**Comment by joein** [comment]
Hey

It can be caused by 2 reasons:

1) you have a very large payload
2) you don't do batching

the former can be solved either by reducing the size of a payload or increasing allowed json size limit
while the latter can be solved by introducing batching

also setting prefer_grpc=True while instantiating a qdrant client would probably solve this problem, but it would be just a way to cure symptoms rather than the problem itself

**Comment by arnavroh45** [comment]
is there a way to increase json payload limit in qdrant db somehow? Because I have tried batching as well, but still getting the same error.

**Comment by joein** [comment]
It is not recommended to increase its size, could you please show how you are trying to upload it?

Which methods do you use?
How do you do batching, what is the batch size?

**Comment by arnavroh45** [comment]
I loaded documents using llama_index document loaders, and then converted them to nodes. 
The problem here I think is, the length/size of the metadata of the nodes(payload for qdrant vectors) is greater than the allocated limit. I implemented the same thing with the documents and it's working but not with the nodes.

sentence_node_parser = SentenceWindowNodeParser.from_defaults(window_size=3, window_metadata_key="window", original_text_metadata_key="original_text")
nodes = sentence_node_parser.get_nodes_from_documents(documents)
vector_store = QdrantVectorStore(client=client, collection_name="collection_name", batch_size=20) #here, specified the batch size
storage_context = StorageContext.from_defaults(vector_store=vector_store)
embeddings = HuggingFaceEmbeddings(model_name='sentence-transformers/all-mpnet-base-v2')
service_context = ServiceContext.from_defaults(llm=llm, embed_model=embeddings, chunk_size=512, chunk_overlap=50)
index = VectorStoreIndex(nodes, storage_context=storage_context, service_context=service_context)

**Comment by joein** [comment]
Could you maybe try it with `batch_size=1`?

If it succeeds, could you check what is actually stored in the payload? Maybe this code stores an original document and not only chunks?

**Comment by arnavroh45** [comment]
Tried that, it isn't working.

**Comment by arnavroh45** [comment]
Did you understand the issue?

**Comment by joein** [comment]
Hi @arnavroh45 , sorry for the delay, haven't had time to look deeper into it yet

@Anush008 maybe you could take a look at it please?

I am not that familiar with llama index, but the error usually occurs when there are either problems with batch size or with payload loading.

**Comment by Anush008** [comment]
Hi @arnavroh45.
I rechecked the [batching implementation](https://github.com/run-llama/llama_index/blob/7cc6b1db27392ff07ecd5d95c5e00a743e212e93/llama_index/vector_stores/qdrant.py#L149). Looks fine.
I'll try reproducing the issue.

**Comment by arnavroh45** [comment]
Okay

**Comment by Anush008** [comment]
Hey @arnavroh45. I tried reproducing this with [SentenceWindowNodeParser](https://docs.llamaindex.ai/en/stable/api/llama_index.node_parser.SentenceWindowNodeParser.html) as per your snippet.
The upload worked fine for me. I assume it has to do with the documents from your data in specific. Could you give some info about the data?

**Comment by arnavroh45** [comment]
I am performing web scraping using selenium and then storing the data in the following format. Then converted the documents into nodes using the provided code. Then adding the nodes to the vector db which gave the error.

Document Format:
Document(id_='dac4c09b-5e50-4c8e-a3de-01ac624ab8e4', embedding=None, metadata={'title': 'title', 'source': 'sourcet'}, excluded_embed_metadata_keys=[], excluded_llm_metadata_keys=[], relationships={}, hash='dd623d76410ed65006c454c4fbf93511baa6bb1bef1936d5a352191b62a0c1bc', text='text', start_char_idx=None, end_char_idx=None, text_template='{metadata_str}\n\n{content}', metadata_template='{key}: {value}', metadata_seperator='\n')

Conversion from document to nodes:
sentence_node_parser = SentenceWindowNodeParser.from_defaults(
    window_size=3,
    window_metadata_key="window",
    original_text_metadata_key="original_text")
nodes = sentence_node_parser.get_nodes_from_documents(documents)

Adding into vector db:
client = QdrantClient(url="url")
vector_store = QdrantVectorStore(client=client, collection_name="collection_name", batch_size=10)
storage_context = StorageContext.from_defaults(vector_store=vector_store)
index = VectorStoreIndex(all_nodes, storage_context=storage_context)

Error:
UnexpectedResponse: Unexpected Response: 400 (Bad Request)
Raw response content:
b'{"status":{"error":"Payload error: JSON payload (41900140 bytes) is larger than allowed (limit: 33554432 bytes)."},"time":0.0}'

**Comment by Anush008** [comment]
@arnavroh45, the document schema is very similar to what I had in my attempt at reproducing.

But this would be almost not possible to debug without the actual data which seems to be the cause, since the batch upload seems to work fine and you even tried setting it to 1 point.

**Comment by arnavroh45** [comment]
Can this be possible that the length of my metadata exceeds the size limit allowed for metadata?

**Comment by arnavroh45** [comment]
Can I get your discord id, so that I can share the data with you.

**Comment by thanhlap0572002** [comment]
I have the same error but I give batch size to 1 but still not work:

from transformers import ViTImageProcessor, ViTModel
from qdrant_client import QdrantClient
from qdrant_client.http import models
from datasets import load_dataset
import numpy as np
import torch

connection = QdrantClient(
    url="https://785dad3b-d933-4697-87ad-93ccebc059d8.us-east4-0.gcp.cloud.qdrant.io:6333",
    api_key="O8iHYXhv7fuR1IW5vA8OLGeFEpQmr6W2lqPfQbUC5iglpcPAgyuG1A",
)
connection.create_collection(
    collection_name="image_similarity",
    vectors_config=models.VectorParams(size=384, distance=models.Distance.COSINE),
)
import os

# Đường dẫn đến thư mục chứa các thư mục hình ảnh
image_root_folder = '/content/drive/MyDrive/qdrant/image_similarity/data'

# Tạo hoặc mở tệp văn bản để lưu nhãn
label_file = 'labels.txt'
with open(label_file, 'w') as f:
    # Lặp qua tất cả các thư mục con trong thư mục gốc
    for label in os.listdir(image_root_folder):
        label_folder = os.path.join(image_root_folder, label)
        if os.path.isdir(label_folder):
            # Lặp qua tất cả các tệp hình ảnh trong thư mục nhãn
            for filename in os.listdir(label_folder):
                image_path = os.path.join(label_folder, filename)
                
                # Lưu tên ảnh và nhãn tương ứng vào tệp văn bản
                f.write(f"{image_path} {label}\n")

processor = ViTImageProcessor.from_pretrained('facebook/dino-vits16')
model = ViTModel.from_pretrained('facebook/dino-vits16')

import cv2
embeddings = []
labels = []

# Lặp qua tất cả các hình ảnh trong thư mục và tính toán embedding
for label in os.listdir(image_root_folder):
    label_folder = os.path.join(image_root_folder, label)
    if os.path.isdir(label_folder):
        for filename in os.listdir(label_folder):
            image_path = os.path.join(label_folder, filename)
            image = cv2.imread(image_path)
            
            # Tiền xử lý hình ảnh
            # Ví dụ: chuyển đổi hình ảnh sang kích thước mà model yêu cầu
            image_resized = cv2.resize(image, (224, 224))  # Ví dụ: resize ảnh về kích thước 224x224
            
            # Chuyển đổi hình ảnh sang tensor và chuẩn hóa
            inputs = processor(images=image_resized, return_tensors="pt", padding=True)
            
            # Feed-forward hình ảnh qua mô hình để tính toán embedding
            outputs = model(**inputs)
            embedding = outputs.last_hidden_state.mean(dim=1).squeeze().detach().numpy()
            
            # Lưu embedding và nhãn tương ứng vào danh sách
            embeddings.append(embedding)
            labels.append(label)

# Chuyển đổi danh sách embedding và nhãn thành định dạng numpy array
embeddings = np.array(embeddings)
labels = np.array(labels)


# Thiết lập batch size load
batch_size = 1000


# Tính toán số lượng batch
num_batches = len(embeddings) // batch_size + (1 if len(embeddings) % batch_size != 0 else 0)

# Lặp qua từng batch
for i in range(num_batches):
    # Tính chỉ số bắt đầu và kết thúc của batch hiện tại
    start_idx = i * batch_size
    end_idx = min((i + 1) * batch_size, len(embeddings))

    # Chia dữ liệu thành batch
    batch_embeddings = embeddings[start_idx:end_idx]
    batch_labels = labels[start_idx:end_idx]

    # Tạo payload cho từng vector embedding
    batch_payloads = []
    for emb, label in zip(batch_embeddings, batch_labels):
        batch_payloads.append({
            "id": str(label),  # Định danh của mỗi vector embedding
            "vector": emb.tolist(),  # Vector embedding chuyển đổi sang định dạng list
            "metadata": {"label": label}  # Thêm metadata cho mỗi vector embedding
        })

    # Gửi yêu cầu upsert cho batch hiện tại lên Qdrant
    connection.upsert(
        collection_name="image_similarity",
        points=models.Batch(
            ids=batch_labels,  # Sử dụng nhãn làm IDs
            vectors=batch_embeddings,
            payloads=batch_payloads
        )
    )

**Comment by thanhlap0572002** [comment]
[erorr.txt](https://github.com/qdrant/qdrant-client/files/14522601/erorr.txt)
i have the same erorr :((

**Comment by Anush008** [comment]
Hi @thanhlap0572002.
Can you try updating to the latest LlamaIndex version?
A PR was merged that updates the `batch_size` implementation. Also, could you try increasing your `batch_size` value?

**Comment by thanhlap0572002** [comment]
I tried increasing batch_size but it still doesn't work

**Comment by thanhlap0572002** [comment]
can u fix for me in discord pls . my id : helpmeplsomgn

**Comment by Anush008** [comment]
Are you on the latest LlamaIndex version? 

If you could you provide a reproducible script?

**Comment by thanhlap0572002** [comment]
`from transformers import ViTImageProcessor, ViTModel
from qdrant_client import QdrantClient
from qdrant_client.http import models
from datasets import load_dataset
import numpy as np
import torch

connection = QdrantClient(
    url="httpslllllllllllllllllllllllllllllllllllllllll33",
    api_key="Olllllllllllllllllllllllllll",
)
connection.create_collection(
    collection_name="image_similarity",
    vectors_config=models.VectorParams(size=384, distance=models.Distance.COSINE),
)
import os

# Đường dẫn đến thư mục chứa các thư mục hình ảnh
image_root_folder = '/content/drive/MyDrive/qdrant/image_similarity/data'

# Tạo hoặc mở tệp văn bản để lưu nhãn
label_file = 'labels.txt'
with open(label_file, 'w') as f:
    # Lặp qua tất cả các thư mục con trong thư mục gốc
    for label in os.listdir(image_root_folder):
        label_folder = os.path.join(image_root_folder, label)
        if os.path.isdir(label_folder):
            # Lặp qua tất cả các tệp hình ảnh trong thư mục nhãn
            for filename in os.listdir(label_folder):
                image_path = os.path.join(label_folder, filename)
                
                # Lưu tên ảnh và nhãn tương ứng vào tệp văn bản
                f.write(f"{image_path} {label}\n")

processor = ViTImageProcessor.from_pretrained('facebook/dino-vits16')
model = ViTModel.from_pretrained('facebook/dino-vits16')

import cv2
embeddings = []
labels = []

# Lặp qua tất cả các hình ảnh trong thư mục và tính toán embedding
for label in os.listdir(image_root_folder):
    label_folder = os.path.join(image_root_folder, label)
    if os.path.isdir(label_folder):
        for filename in os.listdir(label_folder):
            image_path = os.path.join(label_folder, filename)
            image = cv2.imread(image_path)
            
            # Tiền xử lý hình ảnh
            # Ví dụ: chuyển đổi hình ảnh sang kích thước mà model yêu cầu
            image_resized = cv2.resize(image, (224, 224))  # Ví dụ: resize ảnh về kích thước 224x224
            
            # Chuyển đổi hình ảnh sang tensor và chuẩn hóa
            inputs = processor(images=image_resized, return_tensors="pt", padding=True)
            
            # Feed-forward hình ảnh qua mô hình để tính toán embedding
            outputs = model(**inputs)
            embedding = outputs.last_hidden_state.mean(dim=1).squeeze().detach().numpy()
            
            # Lưu embedding và nhãn tương ứng vào danh sách
            embeddings.append(embedding)
            labels.append(label)

# Chuyển đổi danh sách embedding và nhãn thành định dạng numpy array
embeddings = np.array(embeddings)
labels = np.array(labels)


# Thiết lập batch size load
batch_size = 1000


# Tính toán số lượng batch
num_batches = len(embeddings) // batch_size + (1 if len(embeddings) % batch_size != 0 else 0)

# Lặp qua từng batch
for i in range(num_batches):
    # Tính chỉ số bắt đầu và kết thúc của batch hiện tại
    start_idx = i * batch_size
    end_idx = min((i + 1) * batch_size, len(embeddings))

    # Chia dữ liệu thành batch
    batch_embeddings = embeddings[start_idx:end_idx]
    batch_labels = labels[start_idx:end_idx]

    # Tạo payload cho từng vector embedding
    batch_payloads = []
    for emb, label in zip(batch_embeddings, batch_labels):
        batch_payloads.append({
            "id": str(label),  # Định danh của mỗi vector embedding
            "vector": emb.tolist(),  # Vector embedding chuyển đổi sang định dạng list
            "metadata": {"label": label}  # Thêm metadata cho mỗi vector embedding
        })

    # Gửi yêu cầu upsert cho batch hiện tại lên Qdrant
    connection.upsert(
        collection_name="image_similarity",
        points=models.Batch(
            ids=batch_labels,  # Sử dụng nhãn làm IDs
            vectors=batch_embeddings,
            payloads=batch_payloads
        )
    )
`

**Comment by thanhlap0572002** [comment]
`from transformers import ViTImageProcessor, ViTModel
from qdrant_client import QdrantClient
from qdrant_client.http import models
from datasets import load_dataset
import numpy as np
import torch

connection = QdrantClient(
    url="httpslllllllllllllllllllllllllllllllllllllllll33",
    api_key="Olllllllllllllllllllllllllll",
)
connection.create_collection(
    collection_name="image_similarity",
    vectors_config=models.VectorParams(size=384, distance=models.Distance.COSINE),
)
import os


image_root_folder = '/content/drive/MyDrive/qdrant/image_similarity/data'


label_file = 'labels.txt'
with open(label_file, 'w') as f:

    for label in os.listdir(image_root_folder):
        label_folder = os.path.join(image_root_folder, label)
        if os.path.isdir(label_folder):
          
            for filename in os.listdir(label_folder):
                image_path = os.path.join(label_folder, filename)          
                f.write(f"{image_path} {label}\n")

processor = ViTImageProcessor.from_pretrained('facebook/dino-vits16')
model = ViTModel.from_pretrained('facebook/dino-vits16')

import cv2
embeddings = []
labels = []


for label in os.listdir(image_root_folder):
    label_folder = os.path.join(image_root_folder, label)
    if os.path.isdir(label_folder):
        for filename in os.listdir(label_folder):
            image_path = os.path.join(label_folder, filename)
            image = cv2.imread(image_path)
        
            image_resized = cv2.resize(image, (224, 224))  
            

            inputs = processor(images=image_resized, return_tensors="pt", padding=True)
            
    
            outputs = model(**inputs)
            embedding = outputs.last_hidden_state.mean(dim=1).squeeze().detach().numpy()
            
           
            embeddings.append(embedding)
            labels.append(label)


embeddings = np.array(embeddings)
labels = np.array(labels)



batch_size = 1000


num_batches = len(embeddings) // batch_size + (1 if len(embeddings) % batch_size != 0 else 0)


for i in range(num_batches):

    start_idx = i * batch_size
    end_idx = min((i + 1) * batch_size, len(embeddings))


    batch_embeddings = embeddings[start_idx:end_idx]
    batch_labels = labels[start_idx:end_idx]


    batch_payloads = []
    for emb, label in zip(batch_embeddings, batch_labels):
        batch_payloads.append({
            "id": str(label), 
            "vector": emb.tolist(), 
            "metadata": {"label": label}
        })

    connection.upsert(
        collection_name="image_similarity",
        points=models.Batch(
            ids=batch_labels,  
            vectors=batch_embeddings,
            payloads=batch_payloads
        )
    )
`

**Comment by thanhlap0572002** [comment]
[code.txt](https://github.com/qdrant/qdrant-client/files/14522704/code.txt)

**Comment by thanhlap0572002** [comment]
thank youuu , is my folder like that:
![image](https://github.com/qdrant/qdrant-client/assets/114754250/adf13051-4b3b-41ae-8882-d20b21c1a2aa)
